### PR TITLE
Replace illegal statsd characters.

### PR DESCRIPTION
**Pull request by badcure**
It is possible for JMX to have characters invalid for statsd. This causes to parsing to become incorrect.

The unit test added is what we are seeing, where the scope can have "127.0.0.1:8080" where statsd uses the colon as a delimiter between the key and value.

I added an option "invalidCharValue" so that it can be customized.

### CODE REVIEW

File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/StatsDWriter2.java [hunk 1]
```diff
@@ -63,12 +66,15 @@ public StatsDWriter2(
 			@Nonnull String rootPrefix,
 			@Nonnull String bucketType,
 			boolean stringsValuesAsKey,
-			@Nonnull Long stringValueDefaultCount) {
+			@Nonnull Long stringValueDefaultCount,
+			@Nonnull String invalidCharValue) {
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/StatsDWriter2.java]: You should pass directly the pattern here, and initialize it in the related Factory.

Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/StatsDWriter2.java]: I would probably rename `invalidCharValue` to `replacementChar` or `replacementForInvalidChar` ...

File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/StatsDWriter.java [hunk 2]
```diff
@@ -188,8 +194,15 @@ public void internalWrite(Server server, Query query, ImmutableList<Result> resu
 					continue;
 				}
 
-				String line = KeyUtils.getKeyString(server, query, result, values, typeNames, rootPrefix)
-						+ computeActualValue(values.getValue()) + "|" + bucketType + "\n";
+				String line = KeyUtils.getKeyString(server, query, result, values, typeNames, rootPrefix);
+
+				// These characters can mess with formatting.
+				String replaceString = "_";
+				if (this.invalidCharValue != null){
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/StatsDWriter.java]: Checking the nullity of invalidCharValue should be done in constructor.

File: jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/output/BaseOutputWriter.java [hunk 3]
```diff
@@ -121,6 +121,10 @@ public BaseOutputWriter(
 		return first != null ? first : (second != null ? second : checkNotNull(third));
 	}
 
+	protected <T> T firstNonNull(@Nullable T first, @Nullable T second) {
```
Comment (gehel) [inline jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/output/BaseOutputWriter.java]: Guava's Objects.firstNonNull() already does exactly what this method does.

**Comment by badcure** [comment]
@gehel so looks like App Veyor failed due to the PR not being mergable even though git says it is. Travis-ci passed. 

I made the requested changes, but let me know if there is something I need to do to get appveyor to pass.

File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/StatsDWriter.java [hunk 4]
```diff
@@ -134,7 +134,7 @@ public StatsDWriter(
 		if (port == null) {
 			port = Settings.getIntegerSetting(getSettings(), PORT, null);
 		}
-		this.replacementForInvalidChar = firstNonNull(replacementForInvalidChar, "_");
+		this.replacementForInvalidChar = com.google.common.base.MoreObjects.firstNonNull(replacementForInvalidChar, "_");
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/StatsDWriter.java]: import of MoreObjects would probably be more readable

**Comment by gehel** [comment]
AppVeyor is a check that we've added fairly recently. Not sure why it is failing here. I suspect a rebase of your PR might help, but I don't even see that build in the history (https://ci.appveyor.com/project/gquintana/jmxtrans/history). This might just be a transient failure. If travis is happy, I'll merge (see the last comment first).

**Comment by gehel** [comment]
Thanks for the fix!

**Pull request merged**